tests/test_other.py

```diff
@@ -1524,6 +1524,53 @@ def test(link_flags, lib_suffix):
     test(['-lfile'], '') # -l, auto detection from library path
     test([self.in_dir('libdir', 'libfile.so.3.1.4.1.5.9')], '.3.1.4.1.5.9') # handle libX.so.1.2.3 as well
 
+  def test_dynamic_link_module_heap_in_pthread(self):
+    # Test that a side module uses the same heap for global objects across all threads
+
+    # A side module with a global object with a constructor.
+    # The global object must have a constructor to make sure
+    # verify we construct it only once.
+    create_test_file('side.cpp', r'''
+      struct Data {
+          Data() : value(42) {}
+          int value;
+      } data;
```

Comment:
Can this just be replaced with `static int data = 42`?

Reply:
A static int would be initialized when the memory segment gets initialized, which happens during module instantiation.
An object with a constructor would get initialized when global constructors are called, which is triggered by `__post_instantiate`.
One of the fixes in this change is that `__post_instantiate` is only called once (in the main thread). Using an object with a constructor was a deliberate choice.

Reply:
Ah I see good point.

How would this test detect if the constructor was run twice?

Perhaps we should have the constructor have side effect other than data initilization so that it can't be eliminated completely by an optimizer?  Perhaps it could write do a `+= 1` on the `static int`?    It might also make it easier to debug if it included a `puts("ctor called")` so that you can see in the console if its called twice?

Reply:
Ah, good call.
When I wrote this change, this was true:

The constructor initializes the static int to `42`.
The code in `main` sets the value to `123` before creating the thread.
If creating the thread runs the constructor again, it resets the value to `42`.

However, with #13603, constructors are only called once anyway:
Before #13603, `__post_instantiate` was called immediately on the pthread.
After #13603 the `runtimeInitialized` check changed, `__post_instantiate` gets pushed into `__ATINIT__`. However, the content of `__ATINIT__` is not executed in the pthread (`initRuntime` has an early return), resulting in the constructors being called only once.

So I guess what this change really does now is not zero-initialize the static memory segment twice, and avoid a pointless entry being added to `__ATINIT__` in pthreads.

Reply:
To recap:
* Removing the `if (!ENVIRONMENT_IS_PTHREAD) {` check that guards the call to `__post_instantiate` has no effect now, other than preventing an irrelevant push to `__ATINIT__`.
* Removing the `if (!ENVIRONMENT_IS_PTHREAD) {` check that guards the zero-init of the memory does have an effect, but that would be true even if the test used a `static int` rather than an object with constructor.

I wonder if there's still value in having a test that verifies the constructor runs once or not. It's a happy but AFAIK untested consequence of #13603.

Reply:
Thank you for the explanation.   I think both your fixes are good, I just didn't fully understand what was happening in the test.

Can perhaps add a comment to the test saying what is trying to test for exactly.  Something like:  "Tests that thread loading of dynamic libraries that happen on thread creation does not zero-only data segments or call static constructors".?

Other this I think this is good to go now.

Reply:
I see you already added a comment above which seems fine.